File: src/scripts/createEntitiesAndTypes/createEntitiesAndTypes.py
import os
import csv
import requests
import sys

DEFAULT_MENTORING_DOMAIN = 'http://localhost:3569'
MENTORING_DOMAIN = DEFAULT_MENTORING_DOMAIN

access_token = ''
entity_type_id = ''

def main():
    try:
        prompt_for_domain()
        prompt_for_access_token()

        while True:
            chosen_file = select_csv_file()
            if not chosen_file:
                break

            model_names = prompt_for_model_names()
            entity_type_data = build_entity_type_data(chosen_file, model_names)
            create_entity_type(entity_type_data)

            process_csv_file(chosen_file)

            print('\n----------------------------------------\n')
            print('Choose another CSV file or type "exit" to quit.')
            print('\n----------------------------------------\n')
    except Exception as e:
        print(f"Error occurred: {e}")
    finally:
        sys.stdin.close()

def prompt_for_domain():
    global MENTORING_DOMAIN
    domain = input(f"Enter domain (default: {DEFAULT_MENTORING_DOMAIN}): ")
    if domain:
        MENTORING_DOMAIN = domain
    print(f"Using domain: {MENTORING_DOMAIN}")

def prompt_for_access_token():
    global access_token
    access_token = input("Enter access token: ")

def select_csv_file():
    try:
        files = [f for f in os.listdir('.') if f.endswith('.csv')]
        if not files:
            print("No CSV files found.")
            return None

        print("CSV files available:")
        for index, file in enumerate(files):
            print(f"{index + 1}. {file}")
        print("0. Exit")

        choice = input("Choose a CSV file (enter number or 'exit'): ").strip()
        if choice.lower() == 'exit' or choice == '0':
            return None

        file_index = int(choice) - 1
        if 0 <= file_index < len(files):
            chosen_file = files[file_index]
            print(f"Chosen CSV file: {chosen_file}")
            return chosen_file
        else:
            raise ValueError("Invalid choice.")
    except Exception as e:
        print(f"Error: {e}")
        return None

def prompt_for_model_names():
    while True:
        answer = input("Enter model names (space separated): ")
        model_names = answer.split()
        if model_names:
            print(f"Chosen model names: {model_names}")
            return model_names
        else:
            print("No model names provided. Try again.")

def create_entity_type(entity_type_data):
    global entity_type_id
    try:
        response = requests.post(
            f"{MENTORING_DOMAIN}/mentoring/v1/entity-type/create",
            json=entity_type_data,
            headers={
                'x-auth-token': f"bearer {access_token}",
                'Content-Type': 'application/json'
            }
        )
        response.raise_for_status()
        entity_type_id = response.json()['result']['id']
        print(f"Entity type created successfully. Entity type ID: {entity_type_id}")
    except requests.RequestException as e:
        print(f"Entity type creation failed: {e}")
        raise

def process_csv_file(chosen_file):
    try:
        with open(chosen_file, 'r', newline='', encoding='utf-8') as csvfile:
            reader = csv.DictReader(csvfile)
            for row in reader:
                identifier = row['identifier']
                entity = row['entity']
                create_entity(identifier, entity)
        print("All entities created successfully.")
    except Exception as e:
        print(f"Error processing CSV file: {e}")

def create_entity(identifier, entity, retries=3):
    try:
        response = requests.post(
            f"{MENTORING_DOMAIN}/mentoring/v1/entity/create",
            json={
                'value': identifier.strip(),
                'label': entity.strip(),
                'status': 'ACTIVE',
                'type': 'SYSTEM',
                'entity_type_id': entity_type_id
            },
            headers={
                'x-auth-token': f"bearer {access_token}",
                'Content-Type': 'application/json'
            }
        )
        response.raise_for_status()
        print(f"Entity created successfully: {identifier} - {entity}")
    except requests.RequestException as e:
        if retries > 0:
            print(f"Failed to create entity ({identifier} - {entity}). Retrying... ({retries} retries left)")
            create_entity(identifier, entity, retries - 1)
        else:
            print(f"Max retries reached for ({identifier} - {entity}). Skipping.")

def build_entity_type_data(chosen_file, model_names):
    value = chosen_file.replace('.csv', '').lower()
    label = ' '.join(word.capitalize() for word in value.split('_'))
    return {
        'value': value,
        'label': label,
        'status': 'ACTIVE',
        'allow_filtering': False,
        'data_type': 'STRING',
        'allow_custom_entities': True,
        'model_names': model_names,
    }

if __name__ == '__main__':
    main()

File: src/scripts/createEntitiesAndTypes/createEntitiesAndTypes.js
const readline = require('readline')
const fs = require('fs')
const axios = require('axios')

const rl = readline.createInterface({
	input: process.stdin,
	output: process.stdout,
})

let accessToken = ''
let entityTypeId = ''

const DEFAULT_MENTORING_DOMAIN = 'http://localhost:3569'
let MENTORING_DOMAIN = DEFAULT_MENTORING_DOMAIN

async function main() {
	try {
		await promptForDomain()
		await promptForAccessToken()

		while (true) {
			const chosenFile = await selectCsvFile()
			if (!chosenFile) break

			const modelNames = await promptForModelNames()
			const entityTypeData = buildEntityTypeData(chosenFile, modelNames)
			await createEntityType(entityTypeData)

			await processCsvFile(chosenFile)

			console.log('\n----------------------------------------\n')
			console.log('Choose another CSV file or type "exit" to quit.')
			console.log('\n----------------------------------------\n')
		}
	} catch (error) {
		console.error('Error occurred:', error)
	} finally {
		rl.close()
	}
}

async function promptForDomain() {
	const domain = await promptQuestion(`Enter domain (default: ${DEFAULT_MENTORING_DOMAIN}): `)
	if (domain) {
		MENTORING_DOMAIN = domain
	}
	console.log(`Using domain: ${MENTORING_DOMAIN}`)
}

async function promptForAccessToken() {
	accessToken = await promptQuestion('Enter access token: ')
}

function selectCsvFile() {
	return new Promise((resolve, reject) => {
		fs.readdir(__dirname, (err, files) => {
			if (err) {
				console.error('Error reading directory:', err)
				reject(err)
				return
			}

			const csvFiles = files.filter((file) => file.endsWith('.csv'))
			console.log('CSV files available:')
			csvFiles.forEach((file, index) => {
				console.log(`${index + 1}. ${file}`)
			})
			console.log('0. Exit')

			rl.question('Choose a CSV file (enter number or "exit"): ', (answer) => {
				if (answer.toLowerCase() === 'exit' || answer === '0') {
					resolve(null)
					return
				}

				const fileIndex = parseInt(answer) - 1
				if (fileIndex >= 0 && fileIndex < csvFiles.length) {
					const chosenFile = csvFiles[fileIndex]
					console.log(`Chosen CSV file: ${chosenFile}`)
					resolve(chosenFile)
				} else {
					reject(new Error('Invalid choice.'))
				}
			})
		})
	})
}

function promptForModelNames() {
	return new Promise((resolve, reject) => {
		rl.question('Enter model names (space separated): ', (answer) => {
			const modelNames = answer.split(' ').filter((name) => name.length > 0)
			if (modelNames.length > 0) {
				console.log('Chosen model names:', modelNames)
				resolve(modelNames)
			} else {
				reject(new Error('No model names provided.'))
			}
		})
	})
}

async function createEntityType(entityTypeData) {
	try {
		const response = await axios.post(
			`${MENTORING_DOMAIN}/mentoring/v1/entity-type/create`,
			JSON.stringify(entityTypeData),
			{
				headers: { 'x-auth-token': `bearer ${accessToken}`, 'Content-Type': 'application/json' },
			}
		)
		entityTypeId = response.data.result.id
		console.log('Entity type created successfully. Entity type ID:', entityTypeId)
	} catch (error) {
		console.error('Entity type creation failed:', error)
		throw error
	}
}

async function processCsvFile(chosenFile) {
	try {
		const csvData = fs.readFileSync(`${__dirname}/${chosenFile}`, 'utf8')
		const lines = csvData.trim().split('\n')
		const headers = lines[0].split(',')

		for (let i = 1; i < lines.length; i++) {
			const line = lines[i].split(',')
			const identifier = line[headers.indexOf('identifier')]
			const entity = line[headers.indexOf('entity')]

			await createEntity(identifier, entity)
		}

		console.log('All entities created successfully.')
	} catch (error) {
		console.error('Error processing CSV file:', error)
	}
}

async function createEntity(identifier, entity, retries = 3) {
	try {
		const response = await axios.post(
			`${MENTORING_DOMAIN}/mentoring/v1/entity/create`,
			JSON.stringify({
				value: identifier.trim(),
				label: entity.trim(),
				status: 'ACTIVE',
				type: 'SYSTEM',
				entity_type_id: entityTypeId,
			}),
			{
				headers: { 'x-auth-token': `bearer ${accessToken}`, 'Content-Type': 'application/json' },
			}
		)
		console.log(`Entity created successfully: ${identifier} - ${entity}`)
	} catch (error) {
		if (error.response) {
			console.error(`Failed to create entity (${identifier} - ${entity}):`, error.response.data)
		} else {
			console.error(`Failed to create entity (${identifier} - ${entity}):`, error.message)
		}

		if (retries > 0) {
			console.log(`Retrying (${retries} retries left)...`)
			await createEntity(identifier, entity, retries - 1)
		} else {
			console.log(`Max retries reached for (${identifier} - ${entity}). Skipping.`)
		}
	}
}

function buildEntityTypeData(chosenFile, modelNames) {
	return {
		value: chosenFile.replace('.csv', '').toLowerCase(),
		label: chosenFile
			.replace('.csv', '')
			.replace(/_/g, ' ')
			.toLowerCase()
			.replace(/\b\w/g, (l) => l.toUpperCase()),
		status: 'ACTIVE',
		allow_filtering: false,
		data_type: 'STRING',
		allow_custom_entities: true,
		model_names: modelNames,
	}
}

function promptQuestion(question, hidden = false) {
	return new Promise((resolve) => {
		rl.question(question, (answer) => {
			resolve(answer.trim())
		})

		if (hidden) {
			rl.output.write('\x1B[?25l')
		}
	})
}

main()
